lib/hvec_map.h

```diff
@@ -0,0 +1,335 @@
+            data.emplace_back(std::piecewise_construct_t(), std::forward_as_tuple(std::move(k)),
+                              std::forward_as_tuple(std::forward<VV>(v)...));
+            new_key = true;
+        } else {
+            if ((new_key = erased[idx])) {
+                erased[idx] = 0;
+                const_cast<KEY &>(data[idx].first) = std::move(k);
+                data[idx].second = VAL(std::forward<VV>(v)...);
+            } else {
+                data[idx].second = VAL(std::forward<VV>(v)...);
+            }
+        }
+        return std::make_pair(iterator(*this, idx), new_key);
+    }
+    std::pair<iterator, bool> insert(const value_type &v) {
+        bool new_key = false;
+        size_t idx = hv_insert(&v.first);
+        if (idx >= data.size()) {
+            idx = data.size();
+            data.push_back(v);
+            new_key = true;
+        } else {
+            if ((new_key = erased[idx])) {
+                erased[idx] = 0;
+                const_cast<KEY &>(data[idx].first) = v.first;
+                data[idx].second = v.second;
+            } else {
+                data[idx] = v.second;
+            }
+        }
+        return std::make_pair(iterator(*this, idx), new_key);
+    }
+    iterator erase(iterator it) {
+        erased[it.idx] = 1;
+        // FIXME -- would be better to call dtor here, but that will cause
+        // problems with the vector when it resized or is destroyed.  Could
+        // use raw memory and manual construct instead.
+        const_cast<KEY &>(it->first) = KEY();
+        it->second = VAL();
+        ++it;
+        --inuse;
+        return it;
+    }
+    size_t erase(const KEY &k) {
+        size_t idx = remove(&k);
+        if (idx + 1 == 0) return 0;
+        if (idx < data.size()) {
+            const_cast<KEY &>(data[idx].first) = KEY();
+            data[idx].second = VAL();
+        }
+        return 1;
+    }
+#ifdef DEBUG
+    using hash_vector_base::dump;
+#endif
+
+ private:
+    std::vector<value_type, ALLOC> data;
+    size_t hashfn(const void *a) const override { return hf(*static_cast<const KEY *>(a)); }
+    bool cmpfn(const void *a, const void *b) const override {
+        return eql(*static_cast<const KEY *>(a), *static_cast<const KEY *>(b));
+    }
+    bool cmpfn(const void *a, size_t b) const override {
+        return eql(*static_cast<const KEY *>(a), data[b].first);
+    }
+    const void *getkey(uint32_t i) const override { return &data[i].first; }
+    void *getval(uint32_t i) override { return &data[i].second; }
+    uint32_t limit() override { return data.size(); }
+    void resizedata(size_t sz) override { data.resize(sz); }
+    void moveentry(size_t to, size_t from) override {
+        // can't just assign, as the keys are const -- need to destruct & construct
+        data[to].~value_type();
+        new (&data[to]) value_type(std::move(data[from]));
+    }
+};
+
+// XXX(seth): We use this namespace to hide our get() overloads from ADL. GCC
```

Comment:
Stale comments?

Reply:
Not really stale -- the comment is correct as to why the weird namespace stuff is done (this is just copied from the ordered_map.h code), which Seth originally put in.  There should be a better way of doing this, but I'm not sure what it is.